### PR TITLE
Proposed Design Principles

### PR DESCRIPTION
**Pull request by daniel-utilityapi**
This is the initial draft of a proposed set of specification design principles for the Customer Data specifications.

Relevant issue: https://github.com/carbon-data-specification/Customer-Data/issues/18

View these Design Principles on the maintainer's website: https://daniel-utilityapi.github.io/Customer-Data/specs/design-principles

## Timeline

* 2023-03-26 - Pull request opened
* 2023-03-30 - Discuss at WG1 meeting
* 2023-03-26 through 2023-04-27 - Open for comments on pull request and during meetings
* 2023-05-11 - (if consensus reached) Merge into main

## Approved by

* UtilityAPI - Daniel Roesler (maintainer, pull request author)
* Google - Hallie Cramer
* Apple - Jordan Hughes
* FlexiDAO - Grzegorz Bytniewski and Eloi Fàbrega Ferrer
* ...add more as they come in

### CODE REVIEW

**Review by halliecramer1 — APPROVED** [review]
This looks good to me!

File: website/specs/design-principles.md [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+Our goal is to have our specifications be adopted and used widely to address
+our target [Use Cases]({{ "/use-cases" | relative_url }}). This means we must
+write our specifications in such a way that when implmented will actually be
+useful to the target organizations, communities, regulators, and stakeholders.
+
+Below are some strategies we will be using to ensure our specifications are useful:
+
+### Focus on end users <a id="focus-on-end-users" href="#focus-on-end-users" class="permalink">🔗</a>
+
+Our specifications will define how a server must operate so that a client user
```
Comment (jhughes22) [inline website/specs/design-principles.md:99]: Perhaps not a conversation for this document, but is there a definition here or elsewhere of what we mean by client? e.g. do we mean third party servers in a client role? user devices?

File: website/specs/design-principles.md [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+code for automated test suites, under open source licenses. By being
+transparent with our testing procedures, both end users and server
+implementers can inspect and suggest improvements to the various test
+suites that we will maintain.
+
+---
+
+## Principle 3: Be Realistic <a id="be-realistic" href="#be-realistic" class="permalink">🔗</a>
+
+The carbon accounting and energy industries are a huge, complicated, and
```
Comment (jhughes22) [inline website/specs/design-principles.md:157]: typo: `are huge`

**Review by jhughes22 — APPROVED** [review]
Looks great to me overall! I appreciate that you put some thought into the broader industry issues and regulatory environment here.

**Comment by eloifabrega** [comment]
Reviewed this and looks good to me. Thank you!

**Comment by greg-flexidao** [comment]
The design principles are well-structured and comprehensive. However, I would suggest adding one more design principle: "Be Adaptable." This principle would emphasize the need for the specifications to be flexible and adaptable to changes in the energy and carbon accounting sectors, as well as evolving technologies and requirements. Given the relative inflexibility and low agility of the energy sector software infrastructure once deployed, I believe its an important one.

Here's a proposed outline for this additional suggested principle:

---

## Principle 4: Be Adaptable <a id="be-adaptable" href="#be-adaptable" class="permalink">🔗</a>
The energy and carbon accounting sectors are constantly evolving, with new technologies, methods, and requirements emerging frequently. To ensure the longevity and relevance of our specifications, we must design them to be adaptable and responsive to changes in these sectors.

### Build in extensibility <a id="extensibility" href="#extensibility" class="permalink">🔗</a>
Design our specifications to support extensibility, allowing for the addition of new features and functionality over time without requiring significant changes to existing implementations. This includes using modular designs and providing clear guidance on how extensions can be integrated into the core specifications.

### Encourage feedback and contributions <a id="feedback-contributions" href="#feedback-contributions" class="permalink">🔗</a>
Promote an open and inclusive community that encourages feedback and contributions from a wide range of stakeholders. By actively soliciting input and engaging with different perspectives, we can ensure that our specifications remain relevant and useful in a changing landscape.

### Regularly review and update <a id="review-update" href="#review-update" class="permalink">🔗</a>
Establish a process for regular review and updates to the specifications to ensure they remain aligned with the latest developments in the energy and carbon accounting sectors. This includes monitoring changes in related standards and best practices, as well as incorporating feedback from the user community.

---

Let me know what you think. Thanks!

File: website/specs/design-principles.md [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+Our [specifications]({{ '/specs' | relative_url }}) will be published online for
+free and can be viewed by anyone without registration or payment.
+
+Additionally, each section of specifications and documentation will have permalinks,
+so that users can directly link to the specific part that they want to reference.
+
+### Make ample complementary documentation and materials <a id="ample-documentation" href="#ample-documentation" class="permalink">🔗</a>
+
+By definition, our specifications must be comprehensive and, well, _specific_.
+This unfortunately means that the specifications themselves are complex and a
```
[the remark this answers is not in the document]
Comment (daniel-utilityapi) [reply website/specs/design-principles.md]: Thanks! Resolved in https://github.com/carbon-data-specification/Customer-Data/pull/22/commits/66fd99899a6d6a47fb091ad766ddb2b2a8751e48

File: website/specs/design-principles.md [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+### Make APIs and data structures easily understood by humans <a id="human-readable" href="#human-readable" class="permalink">🔗</a>
+
+Before a new user writes any client integration, they will typically first
+explore the documentation, examples, and tutorials to try to get familiar
+with how the various APIs, data models, and behaviors are structured. We want
+this initial exploring experience to be intuitive and require minimal time to
+get a basic understanding of how things work.
+
+To aide in that effort, we will be preferring simple, easy-to-understand, and
```
[the remark this answers is not in the document]
Comment (daniel-utilityapi) [reply website/specs/design-principles.md]: Thanks! Resolved in https://github.com/carbon-data-specification/Customer-Data/pull/22/commits/198a7395ec06d0197dbdc0cec2928b4f57b42628

File: website/specs/design-principles.md (hunk 2 above)
Comment (daniel-utilityapi) [reply website/specs/design-principles.md]: Thanks! Fixed in https://github.com/carbon-data-specification/Customer-Data/pull/22/commits/1a4484388be3b0fc826abbd270dfcc62f1306fcf

File: website/specs/design-principles.md (hunk 1 above)
Comment (daniel-utilityapi) [reply website/specs/design-principles.md:103]: Yes, clients are defined in the specs (e.g. https://daniel-utilityapi.github.io/Customer-Data/specs/cdsc-wg1-01/#client). In general, we mean anyone who is using the services from the utilities, which could be third parties (in the case of getting server metadata) or utility account holders (in the case of providing consent for data access).

**Comment by daniel-utilityapi** [comment]
@greg-flexidao Thanks for the proposed new section. I agree that it's good to add adaptability as a principle. Fixed in https://github.com/carbon-data-specification/Customer-Data/pull/22/commits/747d9b4f163a2e5bab2cd9d97310a13e05c517fb

**Comment by mcgeeyoung** [comment]
WattCarbon has reviewed this and approves.

**Review by TechMaz — COMMENTED** [review]
This makes sense to me, thanks!

**Review by steven-ohmconnect — COMMENTED** [review]
This makes sense to me, thanks!

**Pull request merged**